services/web/server/src/simcore_service_webserver/projects/projects_api.py

```diff
@@ -477,32 +528,213 @@ async def trigger_connected_service_retrieve(
+            )
+            with managed_resource(user_id, client_session_id, app) as rt:
+                user_session_id_list: List[
+                    Tuple[int, str]
+                ] = await rt.find_users_of_resource(PROJECT_ID_KEY, project_uuid)
+
+                if not user_session_id_list:
+                    # no one has the project so we lock it
+                    await rt.add(PROJECT_ID_KEY, project_uuid)
+                    return True
+
+                set_user_ids = {uid for uid, _ in user_session_id_list}
+                if set_user_ids.issubset({user_id}):
+                    # we are the only user
+                    if not await _user_has_another_client_open(
+                        user_session_id_list, app
+                    ):
+                        # steal the project
+                        await rt.add(PROJECT_ID_KEY, project_uuid)
+                        await _clean_user_disconnected_clients(
+                            user_session_id_list, app
+                        )
+                        return True
+            return False
+
+    except aioredlock.LockError:
+        # the project is currently locked
+        return False
+
+
+async def try_close_project_for_user(
+    user_id: int,
+    project_uuid: str,
+    client_session_id: str,
+    app: web.Application,
+):
+    with managed_resource(user_id, client_session_id, app) as rt:
+        user_to_session_ids: List[Tuple(int, str)] = await rt.find_users_of_resource(
+            PROJECT_ID_KEY, project_uuid
+        )
+        # first check we have it opened now
+        if (user_id, client_session_id) not in user_to_session_ids:
+            # nothing to do the project is already closed
+            log.warning(
+                "project [%s] is already closed for user [%s].",
+                project_uuid,
+                user_id,
+            )
+            return
+        # remove the project from our list of opened ones
+        log.debug(
+            "removing project [%s] from user [%s] resources", project_uuid, user_id
+        )
+        await rt.remove(PROJECT_ID_KEY)
+    # check it is not opened by someone else
+    user_to_session_ids.remove((user_id, client_session_id))
+    log.debug("remaining user_to_session_ids: %s", user_to_session_ids)
+    if not user_to_session_ids:
+        # NOTE: depending on the garbage collector speed, it might already be removing it
+        fire_and_forget_task(
+            remove_project_interactive_services(user_id, project_uuid, app)
+        )
+    else:
+        log.warning(
+            "project [%s] is used by other users: [%s]. This should not be possible",
+            project_uuid,
+            {uid for uid, _ in user_to_session_ids},
+        )
+
+
 async def _get_project_lock_state(
-    user_id: int, project_uuid: str, app: web.Application
+    user_id: int,
+    project_uuid: str,
+    app: web.Application,
 ) -> ProjectLocked:
+    """returns the lock state of a project
```

Comment:
why not create a test of any of these entries?

Reply:
this is tested in integration tests at the moment.